### ISSUE TITLE
[CURATOR-486] InterProcessMutex.release fails to remove a lock on retries.

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#1006]
<p>We were planning to build a release lock retry, looks like InterProcessMutex.release don't support it.<br/>
It removes currentThread entry from threadData, even if there is an issue releasing lock.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeHeader panelHeader" style="border-bottom-width: 1px;"><b>InterProcessMutex.java</b></div><div class="codeContent panelContent">
<pre class="code-java">
   @Override
    <span class="code-keyword">public</span> void release() <span class="code-keyword">throws</span> Exception
    {
/*
    Note on concurrency: a given lockData instance
    can be only acted on by a single thread so locking isn't necessary
 */

<span class="code-object">Thread</span>      currentThread = <span class="code-object">Thread</span>.currentThread();
LockData    lockData = threadData.get(currentThread);
<span class="code-keyword">if</span> ( lockData == <span class="code-keyword">null</span> )
{
    <span class="code-keyword">throw</span> <span class="code-keyword">new</span> IllegalMonitorStateException(<span class="code-quote">"You <span class="code-keyword">do</span> not own the lock: "</span> + basePath);
}

<span class="code-object">int</span> newLockCount = lockData.lockCount.decrementAndGet();
<span class="code-keyword">if</span> ( newLockCount > 0 )
{
    <span class="code-keyword">return</span>;
}
<span class="code-keyword">if</span> ( newLockCount < 0 )
{
    <span class="code-keyword">throw</span> <span class="code-keyword">new</span> IllegalMonitorStateException(<span class="code-quote">"Lock count has gone negative <span class="code-keyword">for</span> lock: "</span> + basePath);
}
<span class="code-keyword">try</span>
{
    internals.releaseLock(lockData.lockPath);
}
<span class="code-keyword">finally</span>
{
    threadData.remove(currentThread);
}
    }
</pre>
</div></div>

---
<details><summary><i>Originally reported by <a title="Purshotam Shah" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=puru">puru</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-486" target="_blank">InterProcessMutex.release fails to remove a lock on retries.</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jordan Zimmerman" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Not A Problem
<li><b>resolved</b>: 2023-05-10T16:48:30+01:00
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kezhuw">kezhuw</a>:</i>
<p>Curator uses <tt>Guaranteeable.guaranteed</tt> in deleting lock path. In case of exception, it will try the deletion in background. So the result is that:</p>
<ul>
	<li>If session expired, the lock path will be deleted by zk.</li>
	<li>If session not expired, curator will delete it finally in background.<br/>
Basically, <tt>InterProcessMutex.release</tt> don't supposed to be retried. I rarely saw retry-able lock release in case of error.</li>
</ul>
